javalib/src/main/scala/scala/scalanative/nio/fs/FileHelpers.scala

```diff
@@ -0,0 +1,106 @@
+package scala.scalanative.nio.fs
+
+import scalanative.native._
+import scalanative.posix.dirent._
+import scalanative.posix.{errno => e, fcntl, unistd}, e._, unistd.access
+import scalanative.native._, stdlib._, stdio._, string._
+import scala.collection.mutable.UnrolledBuffer
+import scala.reflect.ClassTag
+import java.io.{File, IOException}
+import fcntl._
+
+object FileHelpers {
+  case class Dirent(name: String, tpe: CShort)
+  def list[T: ClassTag](path: String,
+                        f: (String, CShort) => T,
+                        allowEmpty: Boolean = false): Array[T] = Zone {
+    implicit z =>
+      val dir = opendir(toCString(path))
+
+      if (dir == null) {
+        if (!allowEmpty) throw UnixException(path, errno.errno)
+        null
+      } else {
+        val buffer = UnrolledBuffer.empty[T]
+        Zone { implicit z =>
+          var elem = alloc[dirent]
+          var res  = 0
+          while ({ res = readdir(dir, elem); res == 0 }) {
+            val name = fromCString(elem._2.asInstanceOf[CString])
+
+            // java doesn't list '.' and '..', we filter them out.
+            if (name != "." && name != "..") {
+              buffer += f(name, !elem._3)
+            }
+          }
+          closedir(dir)
+          res match {
+            case e if e == EBADF || e == EFAULT || e == EIO =>
+              throw UnixException(path, res)
+            case _ => buffer.toArray
+          }
+        }
+      }
+  }
+
+  def createNewFile(path: String): Boolean =
+    if (path.isEmpty) {
+      throw new IOException(s"No such file or directory $path")
+    } else {
+      Zone { implicit z =>
+        // 438 == 0666
+        open(toCString(path), O_RDWR | O_CREAT | O_EXCL, 438) match {
+          case fd if fd < 0 =>
+            val e = errno.errno
+            if (e != EEXIST) throw UnixException(path, e)
+            false
+          case fd =>
+            if (close(fd) == -1) throw UnixException(path, errno.errno)
+            true
+        }
+      }
+    }
+
+  def createTempFile(prefix: String,
+                     suffix: String,
+                     dir: File,
+                     minLength: Boolean): File =
+    if (prefix == null) throw new NullPointerException
+    else if (minLength && prefix.length < 3)
+      throw new IllegalArgumentException("Prefix string too short")
+    else {
+      val tmpDir       = Option(dir).fold(tempDir)(_.toString)
+      val newSuffix    = Option(suffix).getOrElse(".tmp")
+      var result: File = null
+      do {
+        result = genTempFile(prefix, newSuffix, tmpDir)
+      } while (!createNewFile(result.toString))
+      result
+    }
+
+  def exists(path: String): Boolean =
+    Zone { implicit z =>
+      access(toCString(path), fcntl.F_OK) == 0
+    }
+  private[this] lazy val random = new scala.util.Random()
```

Comment:
We tend to put all the state early in the object/class.

Reply:
Done.